src/entity.js

```diff
@@ -0,0 +1,198 @@
+import extendWith from './util/extend-with';
+import {
+    isArray,
+    isNonNullObject,
+    isNullish,
+    isRecord,
+    isString
+} from './util/type-guard';
+
+export * from './action';
+export * from './embedded-link';
+export * from './link';
+
+export class Entity {
+    #actions;
+    #class;
+    #entities;
+    #links;
+    #properties;
+    #title;
+
+    constructor(options = {}) {
+        const {
+            actions,
+            class: entityClass,
+            entities,
+            links,
+            properties,
+            title,
+            ...extensions
+        } = options ?? {};
+
+        this.actions = actions;
+        this.class = entityClass;
+        this.entities = entities;
+        this.links = links;
+        this.properties = properties;
+        this.title = title;
+
+        extendWith(this, extensions);
+    }
+
+    get actions() {
+        return this.#actions;
+    }
+
+    set actions(value) {
+        this.#actions = coerce.toUniqueSubComponents(
+            value,
+            this.actions,
+            Action.isValid,
+            Action.of
+        );
+    }
+
+    get class() {
+        return this.#class;
+    }
+
+    set class(value) {
+        this.#class = coerce.toOptionalStringArray(value, this.class);
+    }
+
+    get entities() {
+        return this.#entities;
+    }
+
+    set entities(value) {
+        this.#entities = coerceSubComponents(
+            value,
+            this.entities,
+            SubEntity.isValid,
+            SubEntity.of
+        );
+    }
+
+    get links() {
+        return this.#links;
+    }
+
+    set links(value) {
+        this.#links = coerceSubComponents(
+            value,
+            this.links,
+            Link.isValid,
+            Link.of
+        );
+    }
+
+    get properties() {
+        return this.#properties;
+    }
+
+    set properties(value) {
+        if (isRecord(value)) {
+            this.#properties = value;
+        } else if (isNullish(value)) {
+            this.#properties = undefined;
+        }
+    }
+
+    get title() {
+        return this.#title;
+    }
+
+    set title(value) {
+        this.#title = coerce.toOptionalString(value, this.title);
+    }
+
+    toJSON() {
+        const {
+            actions,
+            class: entityClass,
+            entities,
+            links,
+            properties,
+            title,
+            ...extensions
+        } = this;
+        return {
+            actions,
+            class: entityClass,
+            entities,
+            links,
+            properties,
+            title,
+            ...extensions
+        };
+    }
+}
+
+function coerceSubComponents(value, defaultValue, validator, factory) {
+    if (isArray(value)) {
+        return Object.freeze(value.filter(validator).map(factory));
+    } else if (isNullish(value)) {
+        return undefined;
+    } else {
+        return defaultValue;
+    }
+}
+
+export class EmbeddedRepresentation extends Entity {
```

Comment:
Thoughts on `EmbeddedEntity` instead of the longer `EmbeddedRepresentation`, @SeeSharpCode?

Reply:
I feel like `EmbeddedEntity` aligns with Siren a little better since these would be represented in an array called `entities`.

Reply:
However, if `entities` is supposed to be an array of `SubEntity` and `SubEntity` can be an `EmbeddedLink` or this, then maybe `EmbeddedRepresentation` aligns a little better with what it's representing?

Reply:
Also, should this be in its own file like `EmbeddedLink`?

Reply:
`EmbeddedRepresentation` [comes from the core Siren spec](https://github.com/kevinswiber/siren#embedded-representation), but I'm not sure that's helpful or intuitive. My thought with `EmbeddedEntity` is that it's a little shorter and probably makes more sense because it extends `Entity`.

(I _believe_) It can't be in it's own module since it would create a circular (`Entity` needs `SubEntity`, which needs `EmbeddedRepresentation`, which needs `Entity`).